File: leet_code/cpp/34.在排序数组中查找元素的第一个和最后一个位置.cpp
/*
 * @lc app=leetcode.cn id=34 lang=cpp
 *
 * [34] 在排序数组中查找元素的第一个和最后一个位置
 */
#include "common.h"
// @lc code=start
class Solution {
public:
  vector<int> searchRange(vector<int> &nums, int target) {
    int srt = 0, end = 0;
    int lo = 0, hi = nums.size() - 1;
    while (lo <= hi) {
      int mid = (hi - lo) / 2 + lo;
      if (nums[mid] == target) {
        srt = mid;
        end = mid;
        while (srt >= 0 && nums[srt] == target) {
          srt -= 1;
        }
        while (end < nums.size() && nums[end] == target) {
          end = end + 1;
        }
        return {srt + 1, end - 1};
      } else if (nums[mid] > target) {
        hi = mid - 1;
      } else {
        lo = mid + 1;
      }
    }
    return {-1, -1};
  }
};
// @lc code=end

File: python/34.在排序数组中查找元素的第一个和最后一个位置.py
#
# @lc app=leetcode id=34 lang=python3
#
# [34] Find First and Last Position of Element in Sorted Array
#
# from typing import List
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        def getIndex(nums, idx):
            srt,end = idx, idx
            while srt > -1 and nums[srt] == nums[idx]:
                srt -= 1
            while end < len(nums) and nums[end] == nums[idx]:
                end += 1
            return [srt+1,end-1]
        lo, hi = 0, len(nums)-1
        while lo <= hi:
            mid = lo + (hi-lo)//2
            if nums[mid] == target:
                return getIndex(nums, mid)
            elif nums[mid] > target:
                hi = mid - 1
            else:
                lo = mid + 1
        return [-1,-1]
# if __name__ == "__main__":
#     s = Solution()
#     print(s.searchRange([1], 1))
